controllers/route/route_certificaterequest_controller.go

```diff
@@ -0,0 +1,362 @@
+	}
+	// Is the Private Key malformed?
+	pk, err := cmutilpki.DecodePrivateKeyBytes([]byte(nextPrivateKeyPEM))
+	if err != nil {
+		log.V(5).Error(err, "next private key annotation doesn't contain a valid key")
+		delete(instance.GetAnnotations(), cmapi.IsNextPrivateKeySecretLabelKey)
+		err = c.GetClient().Update(ctx, instance)
+		if err != nil {
+			return c.ManageError(ctx, instance, err)
+		}
+		return c.ManageSuccess(ctx, instance)
+	}
+
+	// We have a private key. Look for matching CertificateRequests
+	certificateRequests := &cmapi.CertificateRequestList{}
+	err = c.GetClient().List(ctx, certificateRequests, client.InNamespace(instance.Namespace))
+	if err != nil {
+		return c.ManageError(ctx, instance, err)
+	}
+	// only consider CertificateRequests that match the private key and are owned by this route
+	var ownedCRs []*cmapi.CertificateRequest
+	for _, cr := range certificateRequests.Items {
+		x509CR, err := x509.ParseCertificateRequest(cr.Spec.Request)
+		if err != nil {
+			continue
+		}
+		matches, err := cmutilpki.PublicKeyMatchesCSR(pk.Public(), x509CR)
+		if matches && err == nil {
+			for _, or := range cr.OwnerReferences {
+				if or.Kind == "Route" && or.Name == instance.Name {
+					ownedCRs = append(ownedCRs, cr.DeepCopy())
+					break
+				}
+			}
+		}
+	}
+	// only consider the most recent CR
+	sort.Slice(ownedCRs, func(i, j int) bool {
+		return ownedCRs[i].CreationTimestamp.Time.After(ownedCRs[j].CreationTimestamp.Time)
+	})
+
+	// Possible states:
+	switch {
+	// no CR exists. Create a CR
+	case len(ownedCRs) == 0:
+		cr, err := c.certificateRequestFromRoute(instance, pk)
+		if err != nil {
+			return c.ManageError(ctx, instance, err)
+		}
+		err = c.GetClient().Create(ctx, cr)
+		if err != nil {
+			return c.ManageError(ctx, instance, err)
+		}
+		return c.ManageSuccess(ctx, instance)
+	// CR has been explicitly denied. Not much we can do here.
+	case cmutil.CertificateRequestIsDenied(ownedCRs[0]):
+		log.Info("Not updating route as Certificate Request has been explicitly denied")
+		c.GetRecorder().Event(
+			instance,
+			corev1.EventTypeWarning,
+			cmapi.CertificateRequestReasonDenied,
+			"Not updating route as Certificate Request has been explicitly denied",
+		)
+		return c.ManageSuccess(ctx, instance)
+
+	// CR is invalid. We can't reconcile the Route either.
+	case cmutil.CertificateRequestHasCondition(ownedCRs[0], cmapi.CertificateRequestCondition{
+		Type:   cmapi.CertificateRequestConditionInvalidRequest,
+		Status: "True",
+	}):
+		log.Info("Not updating route as Certificate Request has been rejected by the Issuer")
+		c.GetRecorder().Event(
+			instance,
+			corev1.EventTypeWarning,
+			cmapi.CertificateRequestReasonDenied,
+			"Not updating route as Certificate Request has been rejected by the Issuer",
+		)
+		return c.ManageSuccess(ctx, instance)
+	// CR is ready. Copy the cert/ca into the route.
+	case cmutil.CertificateRequestHasCondition(ownedCRs[0], cmapi.CertificateRequestCondition{
```

Comment:
What if the DNS name changed on the route while this was being requested? We should at the least check that the DNS of the route matches the request for it to be considered the "latest".